internal/cmd/plugin/hibernate/on.go

```diff
@@ -0,0 +1,415 @@
+
+	// Get the instances to be hibernated
+	managedInstances, primaryInstance, err := resources.GetInstancePods(ctx, clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("could not get cluster pods: %w", err)
+	}
+	if primaryInstance.Name == "" {
+		return nil, fmt.Errorf("no primary instance found, cannot proceed with the hibernation")
+	}
+
+	// Get the PVCs that will be hibernated
+	pvcs, err := resources.GetInstancePVCs(ctx, clusterName, primaryInstance.Name)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get PVCs: %w", err)
+	}
+
+	// Get the serial ID of the primary instance
+	primaryInstanceSerial, err := specs.GetNodeSerial(primaryInstance.ObjectMeta)
+	if err != nil {
+		return nil, fmt.Errorf("could not get the primary node: %w", err)
+	}
+
+	contextLogger := log.FromContext(ctx).WithValues(
+		"clusterName", clusterName,
+		"primaryInstance", primaryInstance.Name)
+
+	return &onCommand{
+		ctx:                   log.IntoContext(ctx, contextLogger),
+		cluster:               &cluster,
+		primaryInstanceSerial: primaryInstanceSerial,
+		managedInstances:      managedInstances,
+		primaryInstance:       primaryInstance,
+		pvcs:                  pvcs,
+		force:                 force,
+		shouldRollback:        false,
+	}, nil
+}
+
+// execute executes the `hibernate on` command
+func (on *onCommand) execute() error {
+	// Check the hibernation preconditions
+	if err := on.checkPreconditionsStep(); err != nil {
+		return err
+	}
+
+	on.printAdvancement("hibernation process starting...")
+
+	if err := on.fenceClusterStep(); err != nil {
+		on.shouldRollback = true
+		return err
+	}
+	defer on.rollbackFenceClusterIfNeeded()
+
+	on.printAdvancement("waiting for the cluster to be fenced")
+
+	if err := on.waitInstancesToBeFencedStep(); err != nil {
+		on.shouldRollback = true
+		return err
+	}
+
+	on.printAdvancement("cluster is now fenced, storing primary pg_controldata output")
+
+	if err := on.annotatePVCStep(); err != nil {
+		on.shouldRollback = true
+		return err
+	}
+	defer on.rollBackAnnotationsIfNeeded()
+
+	on.printAdvancement("PVC annotation complete")
+
+	if err := on.deleteResourcesStep(); err != nil {
+		on.shouldRollback = true
+		return err
+	}
+
+	on.printAdvancement("Hibernation completed")
+	return nil
+}
+
+// checkPreconditionsStep checks if the preconditions for the execution of this step are
+// met or not. If they are not met, it will return an error
+func (on *onCommand) checkPreconditionsStep() error {
+	contextLogger := log.FromContext(on.ctx)
+
+	// We should refuse to hibernate a cluster that was fenced already
+	fencedInstances, err := utils.GetFencedInstances(on.cluster.Annotations)
+	if err != nil {
+		return fmt.Errorf("could not check if cluster is fenced: %v", err)
+	}
+
+	if fencedInstances.Len() > 0 {
+		if on.force {
+			contextLogger.Warning("Continuing hibernation procedure even there are fenced instances")
+		} else {
+			return fmt.Errorf("cannot hibernate a cluster with has fenced instances")
```

Comment:
Isn't this something like: `cannot hibernate a cluster that has fenced instances` ?